# Deep Learning

**Deep learning** is the machine learning technique behind the most exciting capabilities in diverse areas like robotics, natural language processing, image recognition, and artificial intelligence, including the famous AlphaGo. In this notebook, you'll gain hands-on, practical knowledge of how to use deep learning with **Keras 2.0**, the latest version of a cutting-edge library for deep learning in Python.

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

path = 'data/dc22/'

## Coding the forward propagation algorithm

In this exercise, you'll write code to do **forward propagation (prediction)** for your first neural network:

<img src="images/deep_learning01.png" alt="" style="width: 400px;"/>

Each data point is a customer. The first input is how many accounts they have, and the second input is how many children they have. The model will predict how many transactions the user makes in the next year.

The input data has been pre-loaded as `input_data`, and the weights are available in a dictionary called `weights`. The array of weights for the first node in the hidden layer are in `weights['node_0']`, and the array of weights for the second node in the hidden layer are in `weights['node_1']`.

The weights feeding into the output node are available in `weights['output']`.

NumPy will be pre-imported for you as np in all exercises.

In [11]:
input_data = np.array([3, 5])
weights = {'output': np.array([2, 7]), 
           'node_0': np.array([2, 4]), 
           'node_1': np.array([ 4, -5])}

In [9]:
weights['node_0']

array([2, 4])

In [10]:
# Calculate node 0 value: node_0_value (dot product of inputs)
node_0_value = (input_data * weights['node_0']).sum()

# Calculate node 1 value: node_1_value
node_1_value = (input_data * weights['node_1']).sum()

# Put node values into array: hidden_layer_outputs
hidden_layer_outputs = np.array([node_0_value, node_1_value])

# Calculate output: output
output = (hidden_layer_outputs * weights['output']).sum()

# Print output
print(output)

-39


## The Rectified Linear Activation Function

An **"activation function"** is a function applied at each node. It converts the node's input into some output.

The **rectified linear activation function (called ReLU)** has been shown to lead to very high-performance networks. This function takes a single number as an input, returning 0 if the input is negative, and the input if the input is positive.

<img src="images/deep_learning02.png" alt="" style="width: 400px;"/>

Here are some examples:
- `relu(3) = 3`
- `relu(-3) = 0`

In [12]:
def relu(input):
    '''Define your relu activation function here'''
    # Calculate the value for the output of the relu function: output
    output = max(input, 0)
    
    # Return the value just calculated
    return(output)

# Calculate node 0 value: node_0_output
node_0_input = (input_data * weights['node_0']).sum()
node_0_output = relu(node_0_input)

# Calculate node 1 value: node_1_output
node_1_input = (input_data * weights['node_1']).sum()
node_1_output = relu(node_1_input)

# Put node values into array: hidden_layer_outputs
hidden_layer_outputs = np.array([node_0_output, node_1_output])

# Calculate model output (do not apply relu)
model_output = (hidden_layer_outputs * weights['output']).sum()

# Print model output
print(model_output)

52


You predicted 52 transactions. Without this activation function, you would have predicted a negative number! The real power of activation functions will come soon when you start tuning model weights.

## Applying the network to many observations/rows of data

You'll now define a function called `predict_with_network()` which will generate predictions for multiple data observations, which are pre-loaded as `input_data`. As before, `weights` are also pre-loaded. In addition, the `relu()` function you defined in the previous exercise has been pre-loaded.

In [13]:
input_data = [np.array([3, 5]), np.array([ 1, -1]), np.array([0, 0]), np.array([8, 4])]
weights = {'node_0': np.array([2, 4]), 
           'node_1': np.array([ 4, -5]), 
           'output': np.array([2, 7])}

In [14]:
# Define predict_with_network()
def predict_with_network(input_data_row, weights):

    # Calculate node 0 value
    node_0_input = (input_data_row * weights['node_0']).sum()
    node_0_output = relu(node_0_input)

    # Calculate node 1 value
    node_1_input = (input_data_row * weights['node_1']).sum()
    node_1_output = relu(node_1_input)

    # Put node values into array: hidden_layer_outputs
    hidden_layer_outputs = np.array([node_0_output, node_1_output])
    
    # Calculate model output
    input_to_final_layer = (hidden_layer_outputs * weights['output']).sum()
    model_output = relu(input_to_final_layer)
    
    # Return model output
    return(model_output)


# Create empty list to store prediction results
results = []
for input_data_row in input_data:
    # Append prediction to results
    results.append(predict_with_network(input_data_row, weights))

# Print results
print(results)

[52, 63, 0, 148]


## Forward propagation in a deeper network

<img src="images/deep_learning03.png" alt="" style="width: 400px;"/>

You now have a model with 2 hidden layers. The values for an input data point are shown inside the input nodes. The weights are shown on the edges/lines. What prediction would this model make on this data point?

Assume the **activation function** at each node is the **identity function**. That is, each node's output will be the same as its input. So the value of the bottom node in the first hidden layer is -1, and not 0, as it would be if the ReLU activation function was used.

<img src="images/deep_learning04.png" alt="" style="width: 400px;"/>


In [30]:
input_data = np.array([1, 1])
weights = {'node_0': np.array([ 2, 4]), 
           'node_1': np.array([ 4, -5]), 
           'node_2': np.array([ 0, 1]),
           'node_3': np.array([ 1, 1]),
           'output': np.array([ 5, 1])}

In [31]:
node_0_output = (input_data * weights['node_0']).sum()
print(node_0_output)
node_1_output = (input_data * weights['node_1']).sum()
print(node_1_output)
layer_outputs = np.array([node_0_output, node_1_output])
print(layer_outputs)

node_2_output = (layer_outputs * weights['node_2']).sum()
print(node_2_output)
node_3_output = (layer_outputs * weights['node_3']).sum()
print(node_3_output)
layer_outputs = np.array([node_2_output, node_3_output])
print(layer_outputs)

# Put node values into array: hidden_layer_outputs
hidden_layer_outputs = layer_outputs

# Calculate model output (do not apply relu)
model_output = (hidden_layer_outputs * weights['output']).sum()

# Print model output
print(model_output)

6
-1
[ 6 -1]
-1
5
[-1  5]
0


## Multi-layer neural networks

In this exercise, you'll write code to do **forward propagation** for a neural network with 2 hidden layers. Each hidden layer has two nodes. The input data has been preloaded as `input_data`. The nodes in the first hidden layer are called `node_0_0` and `node_0_1`. Their weights are pre-loaded as `weights['node_0_0']` and `weights['node_0_1']` respectively.

The nodes in the second hidden layer are called `node_1_0` and `node_1_1`. Their weights are pre-loaded as `weights['node_1_0']` and `weights['node_1_1']` respectively.

We then create a model output from the hidden nodes using weights pre-loaded as `weights['output']`.

<img src="images/deep_learning05.png" alt="" style="width: 400px;"/>


In [44]:
input_data = np.array([3, 5])
weights = {'node_0_0': np.array([2, 4]),
           'node_0_1': np.array([ 4, -5]),
           'node_1_0': np.array([-1,  2]),
           'node_1_1': np.array([1, 2]),
           'output': np.array([2, 7])}

In [45]:
def predict_with_network(input_data, weights):
    # Calculate node 0 in the first hidden layer
    node_0_0_input = (input_data * weights['node_0_0']).sum()
    node_0_0_output = relu(node_0_0_input)

    # Calculate node 1 in the first hidden layer
    node_0_1_input = (input_data * weights['node_0_1']).sum()
    node_0_1_output = relu(node_0_1_input)

    # Put node values into array: hidden_0_outputs
    hidden_0_outputs = np.array([node_0_0_output, node_0_1_output])
    
    # Calculate node 0 in the second hidden layer
    node_1_0_input = (hidden_0_outputs * weights['node_1_0']).sum()
    node_1_0_output = relu(node_1_0_input)

    # Calculate node 1 in the second hidden layer
    node_1_1_input = (hidden_0_outputs * weights['node_1_1']).sum()
    node_1_1_output = relu(node_1_1_input)

    # Put node values into array: hidden_1_outputs
    hidden_1_outputs = np.array([node_1_0_output, node_1_1_output])

    # Calculate model output: model_output
    model_output = relu((hidden_1_outputs * weights['output']).sum())
    
    # Return model_output
    return(model_output)

output = predict_with_network(input_data, weights)
print(output)

182


## The Need for Optimization

<img src="images/deep_learning06.png" alt="" style="width: 400px;"/>

<img src="images/deep_learning07.png" alt="" style="width: 400px;"/>

<img src="images/deep_learning08.png" alt="" style="width: 400px;"/>

<img src="images/deep_learning09.png" alt="" style="width: 400px;"/>

<img src="images/deep_learning10.png" alt="" style="width: 400px;"/>

<img src="images/deep_learning11.png" alt="" style="width: 400px;"/>


## Calculating model errors

What is the error (predicted - actual) for the following network when the input data is `[3, 2]` and the actual value of the target (what you are trying to predict) is `5`? It may be helpful to get out a pen and piece of paper to calculate these values.

<img src="images/deep_learning12.png" alt="" style="width: 400px;"/>


In [71]:
input_data = np.array([3, 2])
weights = {'node_0': np.array([ 2, 1]), 
           'node_1': np.array([0, 0]),
           'output': np.array([ 2, 2])}

In [74]:
# Calculate node 0 value: node_0_output
node_0_input = (input_data * weights['node_0']).sum()
node_0_output = relu(node_0_input)

# Calculate node 1 value: node_1_output
node_1_input = (input_data * weights['node_1']).sum()
node_1_output = relu(node_1_input)

# Put node values into array: hidden_layer_outputs
hidden_layer_outputs = np.array([node_0_output, node_1_output])
print(hidden_layer_outputs)

# Calculate model output (do not apply relu)
model_output = (hidden_layer_outputs * weights['output']).sum()

# Print model output
print(model_output)

[8 0]
16


In [73]:
error = model_output - 5
error

11

## Understanding how weights change model accuracy

Imagine you have to make a prediction for a single data point. The actual value of the target is `7`. The weight going from `node_0` to the output is `2`, as shown below. If you increased it slightly, changing it to `2.01`, would the predictions become more accurate, less accurate, or stay the same?

<img src="images/deep_learning13.png" alt="" style="width: 400px;"/>


In [79]:
input_data = np.array([3, 2])
weights = {'node_0': np.array([ 2, 1]), 
           'node_1': np.array([0, 0]),
           'output': np.array([ 2.01, 2])}

In [80]:
# Calculate node 0 value: node_0_output
node_0_input = (input_data * weights['node_0']).sum()
node_0_output = relu(node_0_input)

# Calculate node 1 value: node_1_output
node_1_input = (input_data * weights['node_1']).sum()
node_1_output = relu(node_1_input)

# Put node values into array: hidden_layer_outputs
hidden_layer_outputs = np.array([node_0_output, node_1_output])
print(hidden_layer_outputs)

# Calculate model output (do not apply relu)
model_output = (hidden_layer_outputs * weights['output']).sum()

# Print model output
print(model_output)

[8 0]
16.08


In [81]:
error = model_output - 7
error

9.079999999999998

Increasing the weight to 2.01 would increase the resulting error from 9 to 9.08, making the predictions less accurate.

## Coding how weight changes affect accuracy

Now you'll get to change `weights` in a real network and see how they affect **model accuracy**!

Have a look at the following neural network:

<img src="images/deep_learning14.png" alt="" style="width: 400px;"/>

Its weights have been pre-loaded as `weights_0`. Your task in this exercise is to update a single weight in `weights_0` to create `weights_1`, which gives a perfect prediction (in which the predicted value is equal to `target_actual`: 3).

Use a pen and paper if necessary to experiment with different combinations. You'll use the `predict_with_network2()` function, which takes an array of data as the first argument, and weights as the second argument.

In [83]:
def predict_with_network2(input_data, weights):
    # Calculate node 0 value: node_0_output
    node_0_input = (input_data * weights['node_0']).sum()
    node_0_output = relu(node_0_input)

    # Calculate node 1 value: node_1_output
    node_1_input = (input_data * weights['node_1']).sum()
    node_1_output = relu(node_1_input)

    # Put node values into array: hidden_layer_outputs
    hidden_layer_outputs = np.array([node_0_output, node_1_output])
    print(hidden_layer_outputs)

    # Calculate model output (do not apply relu)
    model_output = (hidden_layer_outputs * weights['output']).sum()

    # Return model output
    return model_output

In [84]:
# The data point you will make a prediction for
input_data = np.array([0, 3])

# Sample weights
weights_0 = {'node_0': [2, 1],
             'node_1': [1, 2],
             'output': [1, 1]
            }

# The actual target value, used to calculate the error
target_actual = 3

# Make prediction using original weights
model_output_0 = predict_with_network2(input_data, weights_0)

# Calculate error: error_0
error_0 = model_output_0 - target_actual

# Create weights that cause the network to make perfect prediction (3): weights_1
weights_1 = {'node_0': [2, 1],
             'node_1': [1, 2],
             'output': [1, 0]
            }

# Make prediction using new weights: model_output_1
model_output_1 = predict_with_network2(input_data, weights_1)

# Calculate error: error_1
error_1 = model_output_1 - target_actual

# Print error_0 and error_1
print(error_0)
print(error_1)

[3 6]
[3 6]
6
0


## Scaling up to multiple data points

You've seen how different weights will have different accuracies on a single prediction. But usually, you'll want to measure model accuracy on many points. You'll now write code to compare model accuracies for two different sets of weights, which have been stored as `weights_0` and `weights_1`.

`input_data` is a list of arrays. Each item in that list contains the data to make a single prediction. `target_actuals` is a list of numbers. Each item in that list is the actual value we are trying to predict.

In this exercise, you'll use the `mean_squared_error()` function from sklearn.metrics. It takes the true values and the predicted values as arguments.

You'll also use the preloaded `predict_with_network()` function, which takes an array of data as the first argument, and weights as the second argument.

In [89]:
input_data = [np.array([0, 3]), 
              np.array([1, 2]), 
              np.array([-1, -2]), 
              np.array([4, 0])]
weights_0 = {'node_0': np.array([2, 1]), 
             'node_1': np.array([1, 2]), 
             'output': np.array([1, 1])}
weights_1 = {'node_0': np.array([2, 1]),
             'node_1': np.array([1., 1.5]),
             'output': np.array([1., 1.5])}
target_actuals = [1, 3, 5, 7]

In [90]:
from sklearn.metrics import mean_squared_error

# Create model_output_0 
model_output_0 = []
# Create model_output_1
model_output_1 = []

# Loop over input_data
for row in input_data:
    # Append prediction to model_output_0
    model_output_0.append(predict_with_network2(row, weights_0))
    
    # Append prediction to model_output_1
    model_output_1.append(predict_with_network2(row, weights_1))

# Calculate the mean squared error for model_output_0: mse_0
mse_0 = mean_squared_error(target_actuals, model_output_0)

# Calculate the mean squared error for model_output_1: mse_1
mse_1 = mean_squared_error(target_actuals, model_output_1)

# Print mse_0 and mse_1
print("Mean squared error with weights_0: %f" %mse_0)
print("Mean squared error with weights_1: %f" %mse_1)

[3 6]
[3.  4.5]
[4 5]
[4. 4.]
[0 0]
[0 0]
[8 4]
[8. 4.]
Mean squared error with weights_0: 37.500000
Mean squared error with weights_1: 49.890625
